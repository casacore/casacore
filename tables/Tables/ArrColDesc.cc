//# ArrColDesc.cc: Templated class to describe columns of arrays in tables
//# Copyright (C) 1994,1995,1996,1997,1999,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: casa-feedback@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA

#include <casacore/tables/Tables/ArrColDesc.h>
#include <casacore/tables/Tables/ArrColData.h>
#include <casacore/casa/Arrays/IPosition.h>
#include <casacore/casa/IO/AipsIO.h>
#include <casacore/casa/iostream.h>


namespace casacore { //# NAMESPACE CASACORE - BEGIN

ArrayColumnDescBase::ArrayColumnDescBase (const String& name,
                                          const String& comment,
                                          const String& dataManagerType,
                                          const String& dataManagerGroup,
                                          DataType dt, const String& dataTypeId,
                                          Int options,
                                          uInt ndim, const IPosition& shape)
  : BaseColumnDesc (name, comment, dataManagerType, dataManagerGroup,
                    dt, dataTypeId, options, ndim, shape,
                    False, True, False)
{
    if (nrdim_p <= 0) {
	nrdim_p = -1;
    }
}

ArrayColumnDescBase::ArrayColumnDescBase (const ArrayColumnDescBase& that)
: BaseColumnDesc (that)
{}

ArrayColumnDescBase::~ArrayColumnDescBase()
{}

ArrayColumnDescBase& ArrayColumnDescBase::operator=
                                       (const ArrayColumnDescBase& that)
{
    BaseColumnDesc::operator= (that);
    return *this;
}

//# Return the class name.
String ArrayColumnDescBase::className() const
    { return "ArrayColumnDesc<" + dataTypeId(); }

//# Put the object.
//# The data is read by the ctor taking AipsIO.
//# It was felt that putstart takes too much space, so therefore
//# the version is put "manually".
void ArrayColumnDescBase::putDesc (AipsIO& ios) const
{
    ios << (uInt)1;                  // class version 1
    // Formerly a switch was written to determine if a default existed.
    // This switch was always false.
    // Keep writing this switch (which is not used anymore).
    ios << False;
}

void ArrayColumnDescBase::getDesc (AipsIO& ios)
{
    uInt version;
    ios >> version;
    // Formerly a switch was written to determine if a default existed.
    // This switch was always false.
    // So read it in and do not do anything with it.
    Bool sw;
    ios >> sw;
}

//# Show the column.
void ArrayColumnDescBase::show (ostream& os) const
{
    os << "   Name=" << name();
    os << "   DataType=" << dataType();
    if (dataType() == TpOther) {
	os << ", " << dataTypeId();
    }
    if (maxLength() > 0) {
	os << "   MaxLength=" << maxLength();
    }
    os << "   Nrdim=" << ndim();
    os << "   Shape=" << shape();
    os << endl;
    os << "   DataManager=" << dataManagerType();
    os << "/" << dataManagerGroup();
    os << endl;
    os << "   Comment = " << comment() << endl;
}


//# Create a column object from the description.
PlainColumn* ArrayColumnDescBase::makeColumn (ColumnSet* csp) const
{
    PlainColumn* bcp = new ArrayColumnData (this, csp);
    return bcp;
}


} //# NAMESPACE CASACORE - END
