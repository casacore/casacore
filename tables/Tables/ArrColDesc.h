//# ArrColDesc.h: Templated class to describe columns of arrays in tables
//# Copyright (C) 1994,1995,1996,1997,1999,2000
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef TABLES_ARRCOLDESC_H
#define TABLES_ARRCOLDESC_H

//# Includes
#include <casa/aips.h>
#include <tables/Tables/BaseColDesc.h>
#include <casa/Containers/SimOrdMap.h>
#include <casa/Arrays/IPosition.h>


namespace casa { //# NAMESPACE CASA - BEGIN

//# Forward Declarations
class PlainColumn;
class ColumnSet;
template<class T> class Array;

// <summary>
// Templated class for description of table array columns
// </summary>

// <use visibility=export>

// <reviewed reviewer="Gareth Hunt" date="94Nov17" tests="">
// </reviewed>

// <prerequisite>
//   <li> BaseColumnDesc (and its prerequisites)
//   <li> TableDesc
// </prerequisite>

// <etymology>
// This class builds descriptions of table columns where each cell (which
// may also be called a row) will hold an array.
// </etymology>

// <synopsis> 
// ArrayColumnDesc is a templated class for defining a table column
// containing arrays.
//
// The table values are handled by a data manager. This can be
// a storage manager to store the values in a file or it can be
// a virtual column engine to calculate them on-the-fly.
// Only the basic data types are allowed when storing in a file. These are:
//  Bool, uChar, Short, uShort, Int, uInt, float, double,
//  Complex, DComplex and String.
//
// At table creation time (when a table gets created from a table
// description), each column needs to be bound to a data manager.
// If not done explicitly, the table system will bind a column to the
// default manager defined in the column description.
//
// An array column description consists of the following attributes:
// <ul>
//  <li> Name, which has to be unique and must also be different
//         from possible table keyword names.
//  <li> Data type, which is determined by the template parameter
//         (e.g. ArrayColumnDesc<Int>).
//  <li> A data type id, which tells the unique name of non-standard
//         data types (i.e. for data type == TpOther).
//  <li> Comment, which defaults to the empty string.
//         This serves purely as an informational string for the user.
//  <li> Dimensionality. If given, all arrays in the column need
//         to have that dimensionality.
//  <li> Shape. If given, all arrays in the column need to have
//         that shape.
//  <li> Default data manager, which will be used if a column
//         for a newly created table is not explicitly bound to a
//         datamanager.
//  <li> Data manager group, which serves 2 purposes.
//         Firstly it can be used in class SetupNewTable to bind a group
//         of columns.
//         Secondly, when the default data managers are used, it
//         allows, for example, to have 2 AipsIO storage managers.
//         One for one group of columns and one for another group of columns.
//  <li> Options. These are defined in ColumnDesc.h and can be combined
//       by logically or-ing them.
//       <ol>
//        <li>
//         ColumnDesc::FixedShape says that the arrays in all cells
//         of a column have the same shape. This shape must be defined
//         before a table is created. It does not tell if
//         the array is direct or indirect.
//         A FixedShape array is defined in every cell, while for
//         non-FixedShape arrays a cell can be empty.
//        <li>
//         ColumnDesc::Direct determines if an array is directly
//         stored in the table or if it is stored indirectly in a separate
//         file. Direct arrays enforce the FixedShape option.
//         Usually indirect arrays are only read in on command, while
//         direct arrays are held in memory. So the size of the
//         arrays is an important factor.
//       </ol>
//  <li> Default keyword set, which defaults to an empty set.
//         When a table column gets created from the description, it gets
//         a copy of this keyword set as its initial keyword set.
// </ul>
//
// There are several constructors, which allow the definition of most
// of the above mentioned attributes. Others, like the default keyword
// set, have to be defined explicitly.
//
// This class is derived from BaseColumnDesc, thus the functions
// in there also apply to this class.
//
// Once a column description is set up satisfactorily, it must be added
// to a table description before it can be used by the table system.
// </synopsis>

// <example>
// <srcblock>
//     TableDesc tabDesc("tTableDesc", "1", TableDesc::New);
//
//     // Now define array columns.
//     // This one is indirect and has no dimensionality mentioned yet.
//     // Define the keyword UNIT in it.
//     ArrayColumnDesc<Complex> arr1Column("Arr1", "comment for Arr1");
//     arr1Column.rwKeywordSet().define ("UNIT", "Jy");
//     tabDesc.addColumn (arr1Column);
//
//     // This one is indirect and has 3-dim arrays.
//     tabDesc.addColumn (ArrayColumnDesc<Int>("Arr2",
//                                             "comment for Arr2",
//                                             3));
//     // This one is direct and has 2-dim arrays with axis lengths 4 and 7.
//     tabDesc.addColumn (ArrayColumnDesc<uInt>("Arr3",
//                                              "comment for Arr1",
//        				        IPosition(2,4,7),
// 					        ColumnDesc::Direct));
// </srcblock>
// </example>

// <motivation>
// Several column description classes are needed to allow the user
// to define attributes which are special for each column type.
// For scalars the special attribute is the default value.
// They all have to be templated to support arbitrary data types.
// </motivation>

// <templating arg=T>
//  <li> Default constructor
//  <li> Copy constructor
//  <li> Assignment operator
//  <li> <src>static String dataTypeId();  // (not needed for builtin types)</src>
//       This should return the unique "name" of the class.
// </templating>

//# <todo asof="$DATE:$">
//# A List of bugs, limitations, extensions or planned refinements.
//# </todo>

template<class T>
class ArrayColumnDesc : public BaseColumnDesc
{
friend class ColumnDesc;

public:
    // Construct the column with the given name and dimensionality.
    // The data manager type defaults to the StandardStman storage manager.
    // The data manager group defaults to the data manager type.
    // Ndim <=0 means that the number of dimensions is free and will
    // be defined when creating the table (rows). Ndim>0 means that
    // the arrays in this column must have the given dimensionality.
    // The possible options are defined in ColumnDesc.h.
    explicit ArrayColumnDesc (const String& name, Int ndim = -1,
			      int options = 0);

    // Construct the column with the given name, dimensionality, and comment.
    // The data manager type defaults to the StandardStman storage manager.
    // The data manager group defaults to the data manager type.
    // Ndim <=0 means that the number of dimensions is free and will
    // be defined when creating the table (rows). Ndim>0 means that
    // the arrays in this column must have the given dimensionality.
    // The possible options are defined in ColumnDesc.h.
    ArrayColumnDesc (const String& name, const String& comment,
		     Int ndim = -1, int options = 0);

    // Construct the column with the given name, dimensionality, comment,
    // and default data manager type and group.
    // A blank data manager group defaults to the data manager type.
    // Ndim <=0 means that the number of dimensions is free and will
    // be defined when creating the table (rows). Ndim>0 means that
    // the arrays in this column must have the given dimensionality.
    // The possible options are defined in ColumnDesc.h.
    ArrayColumnDesc (const String& name, const String& comment,
		     const String& dataManName, const String& dataManGroup,
		     Int ndim = -1, int options = 0);

    // Construct the column with the given name and shape.
    // The data manager type defaults to the StandardStman storage manager.
    // The data manager group defaults to the data manager type.
    // The possible options are defined in ColumnDesc.h.
    // This constructor can only be used for FixedShape arrays, because the
    // shape of other arrays can only be set per row.
    ArrayColumnDesc (const String& name,
		     const IPosition& shape, int options = 0);

    // Construct the column with the given name, shape, and comment.
    // The data manager type defaults to the StandardStman storage manager.
    // The data manager group defaults to the data manager type.
    // The possible options are defined in ColumnDesc.h.
    // This constructor can only be used for FixedShape arrays, because the
    // shape of other arrays can only be set per row.
    ArrayColumnDesc (const String& name, const String& comment,
		     const IPosition& shape, int options = 0);

    // Construct the column with the given name, shape, comment,
    // and default data manager type and group.
    // A blank data manager group defaults to the data manager type.
    // The possible options are defined in ColumnDesc.h.
    // This constructor can only be used for FixedShape arrays, because the
    // shape of other arrays can only be set per row.
    // If both ndim and shape are given as > 0, ndim should match the length
    // of shape.
    ArrayColumnDesc (const String& name, const String& comment,
		     const String& dataManName, const String& dataManGroup,
		     const IPosition& shape, int options = 0, int ndim=-1);

    // Copy constructor (copy semantics);
    ArrayColumnDesc (const ArrayColumnDesc<T>&);

    ~ArrayColumnDesc();

    // Assignment (copy semantics);
    ArrayColumnDesc<T>& operator= (const ArrayColumnDesc<T>&);

    // Clone this column description to another.
    BaseColumnDesc* clone() const;

    // Get the name of this class. It is used by the registration process.
    // The template argument gets part of the name.
    String className() const;

    // Create a Column object out of this.
    // This is used by class ColumnSet to construct a table column object.
    virtual PlainColumn* makeColumn (ColumnSet*) const;

    // Show the column.
    void show (ostream& os) const;

public:
    // Construct the object to register its class.
    ArrayColumnDesc
      (SimpleOrderedMap<String, BaseColumnDesc* (*)(const String&)>&);

protected:
    // Create the object from AipsIO (this function is registered).
    static BaseColumnDesc* makeDesc(const String& name);

    // Put the object.
    virtual void putDesc (AipsIO&) const;

    // Get the object.
    virtual void getDesc (AipsIO&);
};



} //# NAMESPACE CASA - END

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <tables/Tables/ArrColDesc.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
