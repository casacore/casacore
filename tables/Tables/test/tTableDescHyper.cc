//# tTableDescHyper.cc: Test program for hypercolumns in TableDesc
//# Copyright (C) 1994,1995,1996,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casacore/tables/Tables/TableDesc.h>
#include <casacore/tables/Tables/ScaColDesc.h>
#include <casacore/tables/Tables/ArrColDesc.h>
#include <casacore/casa/Arrays/Vector.h>
#include <casacore/casa/Arrays/ArrayUtil.h>
#include <casacore/casa/Arrays/ArrayIO.h>
#include <casacore/casa/BasicSL/String.h>
#include <casacore/casa/Exceptions/Error.h>
#include <casacore/casa/iostream.h>

#include <casacore/casa/namespace.h>
// <summary> Test program for hypercolumns in TableDesc </summary>

// This program tests the hypercolumn functionality in class TableDesc.
// The results are written to stdout. The script executing this program,
// compares the results with the reference output file.

void a();
void b();
void excpDesc();

int main (int argc, const char*[])
{
    try {
	a();
	b();
	if (argc < 2) {
	    excpDesc();
	}
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } 
    return 0;                           // exit with success status
}

// First build a description.
void a()
{
    // Build the table description.
    TableDesc td ("tTableDescHyper_tmp", "1", TableDesc::New);
    td.addColumn (ScalarColumnDesc<float> ("Time"));
    td.addColumn (ScalarColumnDesc<double> ("Baseline"));
    td.addColumn (ScalarColumnDesc<Complex> ("DataScalar"));
    td.addColumn (ArrayColumnDesc<float>  ("Pol", IPosition(1,16),
					   ColumnDesc::FixedShape));
    td.addColumn (ArrayColumnDesc<float>  ("Freq", 1));
    td.addColumn (ScalarColumnDesc<String> ("Id"));
    td.addColumn (ArrayColumnDesc<float>  ("Data", 2));
    td.addColumn (ArrayColumnDesc<float>  ("Weight", IPosition(2,16,25),
					   ColumnDesc::FixedShape));
    td.defineHypercolumn ("TSMExample1",
			  4,
			  stringToVector ("Data,Weight"),
			  stringToVector ("Pol,Freq,Baseline,Time"),
			  stringToVector ("Id"));
    td.defineHypercolumn ("TSMExample2",
			  4,
			  stringToVector ("Data"));
    td.defineHypercolumn ("TSMExample3",
			  3,
			  stringToVector ("Data,Weight"),
			  stringToVector ("Pol,Freq,Time"));
    td.defineHypercolumn ("TSMExample4",
			  2,
			  stringToVector ("DataScalar"),
			  stringToVector ("Baseline,Time"));
    td.defineHypercolumn ("TSMExample5",
			  4,
			  stringToVector ("Data,Weight"),
			  stringToVector (",,,"));
    td.defineHypercolumn ("TSMExample6",
			  4,
			  stringToVector ("Data,Weight"),
			  stringToVector ("Pol,,,"));
    td.defineHypercolumn ("TSMExample7",
			  4,
			  stringToVector ("Data,Weight"),
			  stringToVector (",Freq,,"));
    td.defineHypercolumn ("TSMExample8",
			  4,
			  stringToVector ("Data,Weight"),
			  stringToVector (",,Time,"));
    td.defineHypercolumn ("TSMExample9",
			  4,
			  stringToVector ("Data,Weight"),
			  stringToVector (",,,Baseline"));
    td.defineHypercolumn ("TSMExample10",
			  4,
			  stringToVector ("Data,Weight"),
			  stringToVector (",Freq,Time,"));
    td.defineHypercolumn ("TSMExample11",
			  4,
			  stringToVector ("Data,Weight"),
			  stringToVector ("Pol,,,Baseline"));

}

void showHyper (const TableDesc& td, const String& name)
{
    Vector<String> idNames;
    Vector<String> coordNames;
    Vector<String> dataNames;
    uInt ndim = td.hypercolumnDesc (name, dataNames, coordNames, idNames);
    cout << name << ": ndim=" << ndim << endl;
    cout << "  Data=" << dataNames;
    cout << "Coord=" << coordNames;
    cout << "Id=" << idNames << endl;
}

void b()
{
    TableDesc td ("tTableDescHyper_tmp");
    cout << td.isHypercolumn ("TSMExamplea")
	 << td.isHypercolumn ("TSMExample1")
	 << td.isHypercolumn ("TSMExample2")
	 << td.isHypercolumn ("TSMExample3")
	 << td.isHypercolumn ("TSMExample4")
	 << td.isHypercolumn ("TSMExampleb") << endl;
    cout << td.hypercolumnNames() << endl;
    showHyper (td, "TSMExample1");
    showHyper (td, "TSMExample2");
    showHyper (td, "TSMExample3");
    showHyper (td, "TSMExample4");
    showHyper (td, "TSMExample5");
    showHyper (td, "TSMExample6");
    showHyper (td, "TSMExample7");
    showHyper (td, "TSMExample8");
    showHyper (td, "TSMExample9");
    showHyper (td, "TSMExample10");
    showHyper (td, "TSMExample11");
    td.show();
}

void excpDesc()
{
    // Build the table description.
    TableDesc td ("", "1", TableDesc::Scratch);
    td.addColumn (ScalarColumnDesc<float> ("Time"));
    td.addColumn (ScalarColumnDesc<Bool> ("TimeNotNum"));
    td.addColumn (ScalarColumnDesc<Short> ("TimeShort"));
    td.addColumn (ScalarColumnDesc<float> ("Baseline"));
    td.addColumn (ArrayColumnDesc<float>  ("Pol", 1));
    td.addColumn (ArrayColumnDesc<float>  ("Freq", 1));
    td.addColumn (ScalarColumnDesc<float> ("Id"));
    td.addColumn (ArrayColumnDesc<float>  ("Data", 2));
    td.addColumn (ArrayColumnDesc<float>  ("Data0"));
    td.addColumn (ArrayColumnDesc<float>  ("Data1", 1));
    td.addColumn (ArrayColumnDesc<float>  ("Weight", 2));
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector (""));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // no data columns
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      0,
			      stringToVector ("Data,Weight"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // ndim < 1
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // ndim != #coord
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,Timex"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Timex does not exist
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight,Datax"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Datax does not exist
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,Time"),
			      stringToVector ("Idx"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Idx does not exist
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,TimeNotNum"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // TimeNotNum not numeric
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,TimeShort"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Coord short not supported
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,Data"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Coord Data is array > 1-dim
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,Pol"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // coord vectors not at start
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data0,Weight"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Data #dim undefined
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data1,Weight"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Data #dim != Weight #ndim
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data1"),
			      stringToVector ("Pol,Freq,Baseline,Time"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Data #dim != #coordVector
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data"),
			      stringToVector ("Pol,Time,Baseline,Time"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Data #dim != #coordVector
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data"),
			      stringToVector (",,Pol,"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Data #dim != #coordVector
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,Time"),
			      stringToVector ("TimeShort"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Id short not supported
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,Time"),
			      stringToVector ("Data"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Id array not supported
    } 
    try {
	td.defineHypercolumn ("TSMExample",
			      4,
			      stringToVector ("Data,Weight"),
			      stringToVector ("Pol,Freq,Baseline,Time"),
			      stringToVector ("Time"));
    } catch (AipsError x) {
	cout << x.getMesg() << endl;             // Time double used 
    } 
}
