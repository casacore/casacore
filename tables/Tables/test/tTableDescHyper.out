011110
[TSMExample1, TSMExample2, TSMExample3, TSMExample4, TSMExample5, TSMExample6, TSMExample7, TSMExample8, TSMExample9, TSMExample10, TSMExample11]
TSMExample1: ndim=4
  Data=[Data, Weight]Coord=[Pol, Freq, Baseline, Time]Id=[Id]
TSMExample2: ndim=4
  Data=[Data]Coord=[]Id=[]
TSMExample3: ndim=3
  Data=[Data, Weight]Coord=[Pol, Freq, Time]Id=[]
TSMExample4: ndim=2
  Data=[DataScalar]Coord=[Baseline, Time]Id=[]
TSMExample5: ndim=4
  Data=[Data, Weight]Coord=[, , , ]Id=[]
TSMExample6: ndim=4
  Data=[Data, Weight]Coord=[Pol, , , ]Id=[]
TSMExample7: ndim=4
  Data=[Data, Weight]Coord=[, Freq, , ]Id=[]
TSMExample8: ndim=4
  Data=[Data, Weight]Coord=[, , Time, ]Id=[]
TSMExample9: ndim=4
  Data=[Data, Weight]Coord=[, , , Baseline]Id=[]
TSMExample10: ndim=4
  Data=[Data, Weight]Coord=[, Freq, Time, ]Id=[]
TSMExample11: ndim=4
  Data=[Data, Weight]Coord=[Pol, , , Baseline]Id=[]

TableDesc tTableDescHyper_tmp   version 1   (Directory ./)
---------
  Comment: 
  #Keywords = 0
  #Columns  = 8
0  Hypercolumn_TSMExample1 : SUBRECORD
1  Hypercolumn_TSMExample2 : SUBRECORD
2  Hypercolumn_TSMExample3 : SUBRECORD
3  Hypercolumn_TSMExample4 : SUBRECORD
4  Hypercolumn_TSMExample5 : SUBRECORD
5  Hypercolumn_TSMExample6 : SUBRECORD
6  Hypercolumn_TSMExample7 : SUBRECORD
7  Hypercolumn_TSMExample8 : SUBRECORD
8  Hypercolumn_TSMExample9 : SUBRECORD
9  Hypercolumn_TSMExample10 : SUBRECORD
10  Hypercolumn_TSMExample11 : SUBRECORD
   Name=Time   DataType=float
   DataManager=TiledShapeStMan/TSMExample10   Default=0
   Comment = 
   #keywords=0

   Name=Baseline   DataType=double
   DataManager=TiledShapeStMan/TSMExample11   Default=0
   Comment = 
   #keywords=0

   Name=DataScalar   DataType=Complex
   DataManager=TiledColumnStMan/TSMExample4   Default=(0,0)
   Comment = 
   #keywords=0

   Name=Pol   DataType=float   Nrdim=1   Shape=[16]
   DataManager=TiledShapeStMan/TSMExample11
   Comment = 
   #keywords=0

   Name=Freq   DataType=float   Nrdim=1   Shape=[]
   DataManager=TiledShapeStMan/TSMExample10
   Comment = 
   #keywords=0

   Name=Id   DataType=String
   DataManager=TiledShapeStMan/TSMExample1   Default=
   Comment = 
   #keywords=0

   Name=Data   DataType=float   Nrdim=2   Shape=[]
   DataManager=TiledShapeStMan/TSMExample11
   Comment = 
   #keywords=0

   Name=Weight   DataType=float   Nrdim=2   Shape=[16, 25]
   DataManager=TiledShapeStMan/TSMExample11
   Comment = 
   #keywords=0

Invalid description of hypercolumn TSMExample: dataColumnNames is empty
Invalid description of hypercolumn TSMExample: ndim < 1
Invalid description of hypercolumn TSMExample: #coordColumnNames mismatches ndim
Invalid description of hypercolumn TSMExample: coordColumn Timex does not exist
Invalid description of hypercolumn TSMExample: dataColumn Datax does not exist
Invalid description of hypercolumn TSMExample: idColumn Idx does not exist
Invalid description of hypercolumn TSMExample: coordColumn TimeNotNum is not numeric
Invalid description of hypercolumn TSMExample: coordColumn TimeShort: (u)Char and (u)Short not supported
Invalid description of hypercolumn TSMExample: coordColumn Data is not a scalar or vector
Invalid description of hypercolumn TSMExample: coordinate vectors have to describe the first axes
Invalid description of hypercolumn TSMExample: the dimensionality of data column Data0 is undefined
Invalid description of hypercolumn TSMExample: the dimensionality of data column Weight mismatches that of previous data columns
Invalid description of hypercolumn TSMExample: the dimensionality of the data columns mismatches nr of coordinate columns with a vector value
Invalid description of hypercolumn TSMExample: the dimensionality of the data columns mismatches nr of coordinate columns with a vector value
Invalid description of hypercolumn TSMExample: the dimensionality of the data columns mismatches nr of coordinate columns with a vector value
Invalid description of hypercolumn TSMExample: idColumn TimeShort: (u)Char and (u)Short not supported
Invalid description of hypercolumn TSMExample: idColumn Data is not a scalar
Invalid description of hypercolumn TSMExample: column name Time is multiply used
